### DISCUSSION TITLE
Set/override KiCad text variables via CLI option

### DISCUSSION THREAD

**Discussion by cdwilson** [INTI-CMNB/KiBot#241] [Ideas]
I'm working on a design which has multiple variants, and I'd like to be able to set/override `${VARIABLE}` values in the KiCad design files differently for each variant.

I'm currently generating KiBot outputs via a make target similar to the following, which changes the global `variant` value via the `-g` CLI option:

```makefile
VARIANTS=variant_1 variant_2

$(VARIANTS):
	kibot -e $(KICAD_SCHEMATIC_FILE) -b $(KICAD_PCB_FILE) -c $(KIBOT_RELEASE_CFG_FILE) -d $(KIBOT_OUTPUT_DIR) -g variant=$@
```

However, AFAIK there does not seem to be a way to set/override text variables via the CLI.

@set-soft Is this something that you would be open to adding?

**Comment by set-soft** [comment]
Hi @cdwilson !

Text variables can be set using shell commands, so you can simply use a command. You can write their value to a file, or just define environment variables and then use commands to get their value (i.e. `cat` to get a file content, or just `echo $VAR` to read the environment).

**Reply by cdwilson** [reply]
@set-soft is there an example you can point me to with a little more detail on how to get this working?  

For example, I tried setting an environment variable for the docker container via `--env SCHEMATIC_PART_NUMBER="foobar"`, but when I run KiBot in the container, I get the warning below, and the value is unset in KiCad.

```
WARNING:(W082) Unknown text variable `SCHEMATIC_PART_NUMBER` (kibot.gs - gs.py:302)
```

I double checked and the environment variable is being set correctly inside the container:

```
root@3d669e65f74d:/home/root/workdir# echo $SCHEMATIC_PART_NUMBER
foobar

```

**Reply by set-soft** [reply]
> I double checked and the environment variable is being set correctly inside the container:
> 
> ```
> root@3d669e65f74d:/home/root/workdir# echo $SCHEMATIC_PART_NUMBER
> foobar
> ```

You can just use what you tried:

```yaml
  set_text_variables:
    - variable: "Comment4"
      command: "echo $SCHEMATIC_PART_NUMBER"
```

The `command` is passed to a shell, so you can use anything the UNIX shell can execute, like `cat FILE_WITH_THE_VALUE`

**Reply by cdwilson** [reply]
> `command: "echo $SCHEMATIC_PART_NUMBER"`

Got it, this was the piece I was missing.  It works now, thanks!

**Comment by cdwilson** [comment]
FYI, I'm not sure why this didn't occur to me earlier, but I ended up solving this without CLI overrides by just using separate `kibot.yaml` files for each variant.  Each `release_<variant>.kibot.yaml` contains only the variant-specific configuration, and they all import the same `release.kibot.yaml` which contains all the shared configuration.

Then, in the makefile, kibot is called with the correct `release_<variant>.kibot.yaml` for each variant target:

```
VARIANTS=variant_1 variant_2
KIBOT_RELEASE_CFG_FILE=release_$@.kibot.yaml

$(VARIANTS):
	kibot -e $(KICAD_SCHEMATIC_FILE) -b $(KICAD_PCB_FILE) -c $(KIBOT_RELEASE_CFG_FILE) -d $(KIBOT_OUTPUT_DIR)
```
